File: code/Splitter.vb
' In order to compile this code:
' C:\sisula\code>C:\WINDOWS\Microsoft.NET\Framework64\v3.5\vbc.exe /optimize /debug- /target:library /out:Splitter.dll Splitter.vb
Imports System
Imports System.Data.Sql
Imports Microsoft.SqlServer.Server
Imports System.Collections
Imports System.Data.SqlTypes
Imports System.Text.RegularExpressions

Public Class Splitter
    <SqlFunction(FillRowMethodName:="FillRow", IsDeterministic:=True, IsPrecise:=True)> _
    Public Shared Function InitMethod(ByVal row As String, ByVal pattern As String) As IEnumerable
        ' Returns a GroupCollection which implements IEnumerable
        Return Regex.Match(row, pattern, RegexOptions.None).Groups
    End Function
    Public Shared Sub FillRow(ByVal obj As Object, ByRef match As SqlChars)
        Dim group As Group = CType(obj, Group)
		' group.Success can apparently be true even if the match is empty
		' so we check for an empty value instead
        If group.Value = String.Empty Then
        	match = New SqlChars(Nothing)
        Else
        	match = New SqlChars(group.Value)
        End If
    End Sub
End Class

File: code/Splitter.cs
// In order to compile this code:
// C:\sisula\code>C:\WINDOWS\Microsoft.NET\Framework64\v3.5\csc.exe /optimize /debug- /target:library /out:Splitter.dll Splitter.cs
using System;
using System.Data.Sql;
using Microsoft.SqlServer.Server;
using System.Collections;
using System.Data.SqlTypes;
using System.Text.RegularExpressions;

public partial class Splitter {
    [
        Microsoft.SqlServer.Server.SqlFunction (
            SystemDataAccess  = SystemDataAccessKind.None,
            DataAccess        = DataAccessKind.None,
            IsDeterministic   = true,
            IsPrecise         = true,
            FillRowMethodName = "FillRow"
        )
    ]
    // GroupCollection
    public static IEnumerable InitMethod(SqlChars row, SqlString pattern) {
        return Regex.Match(row.ToString(), pattern.ToString(), RegexOptions.None).Groups;
    }
    public static void FillRow(Object fromEnumeration, out SqlChars match) {
        Group group = (Group) fromEnumeration;
        if(group.Value == String.Empty) 
            match = new SqlChars(SqlString.Null);
        else
            match = new SqlChars(group.Value);
    }
}
